runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -1649,6 +1659,140 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
    return cursor;
    }
 
+TR::Instruction *
+J9::Z::PrivateLinkage::buildNoPatchingIPIC(TR::Node *callNode, TR::RegisterDependencyConditions *dependencies, TR::Register *vftReg, intptr_t virtualThunk)
+   {
+   /**
+    * TODO: Right now this function generates all the instruction as we are
+    * fixing the number of PIC slots whereas we can change number of slots at compile
+    * time for regular JIT compiled code. It is not that complicated to do that for
+    * readOnlyCode.
+    * We can create a different data structure to hold the J9Class and method address
+    * size of which can be fixed at compilation time and use that.
+    */
+   OMR::CCData *codeCacheData = cg()->getCodeCache()->manager()->getCodeCacheData();
+   OMR::CCData::index_t index;
+
+   if (!(codeCacheData->put(NULL, sizeof(ccInterfaceData), 16, NULL, index)))
+      {
+      cg()->comp()->failCompilation<TR::CompilationException>("Could not allocate interface dispatch metadata");
+      }
+   ccInterfaceData *ccInterfaceDataAddress = codeCacheData->get<ccInterfaceData>(index);
+
+   // Intialization of CCData
+   ccInterfaceDataAddress->slot1Class  = 0;
+   ccInterfaceDataAddress->slot1Method = 0;
+   ccInterfaceDataAddress->slot2Class  = 0;
+   ccInterfaceDataAddress->slot2Method = 0;
+   ccInterfaceDataAddress->slot3Class  = 0;
+   ccInterfaceDataAddress->slot3Method = 0;
+
+   ccInterfaceDataAddress->cpAddress = reinterpret_cast<intptr_t>(callNode->getSymbolReference()->getOwningMethod(cg()->comp())->constantPool());
+   ccInterfaceDataAddress->cpIndex = (intptr_t) callNode->getSymbolReference()->getCPIndexForVM();
+   ccInterfaceDataAddress->interfaceClass = 0;
+   ccInterfaceDataAddress->iTableIndex = 0;
+   ccInterfaceDataAddress->j2iThunkAddress = virtualThunk;
+   ccInterfaceDataAddress->totalSizeOfPicSlots = (3 /*Total Number of PICSlots*/ * sizeof(ccInterfaceDataAddress->slot1Class)) + (3 * sizeof(ccInterfaceDataAddress->slot1Method));
```

Comment:
Can we use a `static const` inside of `ccInterfaceData` to store the constant `3` and use it here instead? This seems quite error prone as the magic constant gets used several times in this function and we can replace it with a properly named and scoped constant that is defined right in the struct.

Reply:
How about using a macro `NUMBER_OF_PIC_SLOTS_FOR_RX` instead of adding additional field in ccInterfaceData? This CCData is allocated per call site and we will have additional field added to all the call site's CCInterfaceData with information which actually is a constant (ATM number of PIC slots are constant and can not be changed at compilation time).

Reply:
`static const` is allocated once for the entire DLL, not per call site. It is a static variable after all. It will not incur any difference to `sizeof(ccInterfaceData)`.

Reply:
Sorry, over sighted the suggestion and forgot we can have static members inside struct in C++. Added the static const inside the ccInterfaceData struct to reflect number of PIC slots, https://github.com/eclipse/openj9/compare/e2eb175cf4e392b9a013648129c1209bc1790bc8..7e34301ef20a2e2b9b78a0d987f8c0e5eb801d0c.